include/oneapi/dpl/pstl/execution_impl.h

```diff
@@ -29,100 +30,132 @@ namespace dpl
+    using __backend_tag = __par_backend_tag;
+};
 
-template <typename _Tp>
-::std::false_type __lazy_and(_Tp, ::std::false_type)
+template <class _IsVector>
+struct __parallel_tag
 {
-    return ::std::false_type{};
-}
+    using __is_vector = _IsVector;
+    using __backend_tag = __par_backend_tag;
+};
+
+struct __parallel_forward_tag
+{
+    using __is_vector = ::std::false_type;
+    using __backend_tag = __par_backend_tag;
+};
+
+template <class _IsVector, class... _IteratorTypes>
+using __tag_type = ::std::conditional_t<
+    __internal::__is_random_access_iterator_v<_IteratorTypes...>, __parallel_tag<_IsVector>,
+    ::std::conditional_t<__is_forward_iterator_v<_IteratorTypes...>, __parallel_forward_tag, __serial_tag<_IsVector>>>;
 
-template <typename _Tp>
-inline _Tp
-__lazy_and(_Tp __a, ::std::true_type)
+template <class... _IteratorTypes>
+__serial_tag<std::false_type>
+__select_backend(oneapi::dpl::execution::sequenced_policy, _IteratorTypes&&...)
 {
-    return __a;
+    return {};
 }
 
-template <typename _Tp>
-::std::true_type __lazy_or(_Tp, ::std::true_type)
+template <class... _IteratorTypes>
+__serial_tag<__internal::__is_random_access_iterator<_IteratorTypes...>>
+__select_backend(oneapi::dpl::execution::unsequenced_policy, _IteratorTypes&&...)
 {
-    return ::std::true_type{};
+    return {};
 }
 
-template <typename _Tp>
-inline _Tp
-__lazy_or(_Tp __a, ::std::false_type)
+template <class... _IteratorTypes>
+__tag_type<std::false_type, _IteratorTypes...>
+__select_backend(oneapi::dpl::execution::parallel_policy, _IteratorTypes&&...)
 {
-    return __a;
+    return {};
 }
 
-/* policy */
-template <typename Policy>
-struct __policy_traits
+template <class... _IteratorTypes>
+__tag_type<__internal::__is_random_access_iterator<_IteratorTypes...>, _IteratorTypes...>
+__select_backend(oneapi::dpl::execution::parallel_unsequenced_policy, _IteratorTypes&&...)
 {
-};
+    return {};
+}
 
-template <>
-struct __policy_traits<oneapi::dpl::execution::sequenced_policy>
+//----------------------------------------------------------
+// __is_backend_tag_serial, __is_backend_tag_serial_v
+//----------------------------------------------------------
+
+template <class _Tag>
+struct __is_backend_tag_serial : ::std::false_type
 {
-    typedef ::std::false_type __allow_parallel;
-    typedef ::std::false_type __allow_unsequenced;
-    typedef ::std::false_type __allow_vector;
 };
 
-template <>
-struct __policy_traits<oneapi::dpl::execution::unsequenced_policy>
+template <class _IsVector>
+struct __is_backend_tag_serial<__serial_tag<_IsVector>> : ::std::true_type
 {
-    typedef ::std::false_type __allow_parallel;
-    typedef ::std::true_type __allow_unsequenced;
-    typedef ::std::true_type __allow_vector;
 };
 
-template <>
-struct __policy_traits<oneapi::dpl::execution::parallel_policy>
+template <class _Tag>
+inline constexpr bool __is_backend_tag_serial_v = __is_backend_tag_serial<::std::decay_t<_Tag>>::value;
```

Comment:
I already had the similar comment for the different review. It's very weird when the trait and `_v` shortcut might produce different values.

Reply:
I don't quite understood what you mean about different values.
But the idea of `__is_backend_tag_serial_v` that I using it to check tag type at places where I should have only serial tag.
For example, let's find what we have for `__pattern_walk1(` :
```C++
template <class _Tag, class _ExecutionPolicy, class _ForwardIterator, class _Function>
void
__pattern_walk1(_Tag, _ExecutionPolicy&&, _ForwardIterator __first, _ForwardIterator __last, _Function __f) noexcept
{
    static_assert(__is_backend_tag_serial_v<_Tag>);

    __internal::__brick_walk1(__first, __last, __f, typename _Tag::__is_vector{});
}

template <class _ExecutionPolicy, class _ForwardIterator, class _Function>
void
__pattern_walk1(__parallel_forward_tag, _ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last,
                _Function __f)
{
    // ...
}

template <class _IsVector, class _ExecutionPolicy, class _ForwardIterator, class _Function>
void
__pattern_walk1(__parallel_tag<_IsVector>, _ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last,
                _Function __f)
{
    // ...
}
```

So this mean that in the first impl we should have only serial tag as far as we have separate implementations for the other tag types:
` static_assert(__is_backend_tag_serial_v<_Tag>);`